### DISCUSSION TITLE
Is gotext safe to use to translate web requests?

### DISCUSSION THREAD

**Discussion by roma-glushko** [leonelquinteros/gotext#100] [General]
I'm curious if it's safe to call `SetlLanguage()` method to internationalize requests coming from users with different language perferences? For example, the webserver receives two requests in parallel, one has to be in Japanese and another one has to be in German, is there a chance of wrong translations because these two requests interleave each other?

**Comment by leonelquinteros** [comment]
Hi @roma-glushko , `gotext.SetLanguage()` is for package level usage. Package level functions are meant to be in the global scope and are available for convenience. This means that it's not safe for your use case. 

What you need is to use the `Locale` object ( https://pkg.go.dev/github.com/leonelquinteros/gotext#Locale ), so you can instantiate a different context (language/config) on each request scope. 

Hope that helps.

**Reply by roma-glushko** [reply]
Hey @leonelquinteros Leonel 👋 

Thanks for the reply!

> gotext.SetLanguage() is for package level usage. Package level functions are meant to be in the global scope and are available for convenience. This means that it's not safe for your use case.

Yeah, that's how I read it 👀  

> What you need is to use the Locale object ( https://pkg.go.dev/github.com/leonelquinteros/gotext#Locale ), so you can instantiate a different context (language/config) on each request scope.

Are you interested in getting this natively in the library? I assume this is a fairly common case to translate some information in HTTP handlers especially for SSR-based applications.

**Comment by leonelquinteros** [comment]
I'm not sure to follow what you're proposing. What would be the use case? 
Internally, the package uses these Locale objects: https://github.com/leonelquinteros/gotext/blob/master/gotext.go#L30-L45

**Reply by roma-glushko** [reply]
@leonelquinteros I thought it's more work to do to propagate locale in the context object all over the places, but it's basically about this:

- set the needed Locale object into the request context by cookie/headers/etc with a custom middleware
- get Locale back from the context when it's time to translate stuff
- call the `Locals.GetX()` methods to do the translation

I was under the impression that there is no way to pass the specific Locales into `gotext.GetN()` and other methods, but it turned out to be not necessary.

Really curious why popular articles like [this](https://phrase.com/blog/posts/internationalization-i18n-go/) doesn't care about concurrency issues when setting/using `Locale` from the global variable. So it might be worth stating somewhere how to get this stuff done not to shoot into your foot.

**Comment by leonelquinteros** [comment]
I didn't know about that article, I'll take a look later! 

For your use case I don't think the package needs to support it, as it's more like a wrapper to fetch the corresponding Locale when needed. 
The flow would look much like you've described. You can also load all languages/Locales at once into a map and then in the request context only fetch from the map instead of parsing the language file again.

**Reply by roma-glushko** [reply]
Cool, I think this will make the trick. Thank you!

**Comment by roma-glushko** [comment]
Resolving the conversation ✅ 
Thanks again for the help!
